x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowMetadata.java

```diff
@@ -0,0 +1,336 @@
+                if (token == XContentParser.Token.FIELD_NAME) {
+                    fieldName = p.currentName();
+                } else if (token == XContentParser.Token.START_OBJECT) {
+                    patterns.put(fieldName, AutoFollowPattern.PARSER.parse(p, c));
+                } else {
+                    throw new ElasticsearchParseException("unexpected token [" + token + "]");
+                }
+            }
+            return patterns;
+        }, PATTERNS_FIELD);
+        PARSER.declareObject(ConstructingObjectParser.constructorArg(), (p, c) -> {
+            Map<String, List<String>> alreadyFollowedIndexUUIDS = new HashMap<>();
+            String fieldName = null;
+            for (XContentParser.Token token = p.nextToken(); token != XContentParser.Token.END_OBJECT; token = p.nextToken()) {
+                if (token == XContentParser.Token.FIELD_NAME) {
+                    fieldName = p.currentName();
+                } else if (token == XContentParser.Token.START_ARRAY) {
+                    alreadyFollowedIndexUUIDS.put(fieldName, Arrays.asList(XContentUtils.readStringArray(p, false)));
+                } else {
+                    throw new ElasticsearchParseException("unexpected token [" + token + "]");
+                }
+            }
+            return alreadyFollowedIndexUUIDS;
+        }, FOLLOWED_LEADER_INDICES_FIELD);
+    }
+
+    public static AutoFollowMetadata fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    private final Map<String, AutoFollowPattern> patterns;
+    private final Map<String, List<String>> followedLeaderIndexUUIDS;
+
+    public AutoFollowMetadata(Map<String, AutoFollowPattern> patterns, Map<String, List<String>> followedLeaderIndexUUIDS) {
+        this.patterns = patterns;
+        this.followedLeaderIndexUUIDS = followedLeaderIndexUUIDS;
+    }
+
+    public AutoFollowMetadata(StreamInput in) throws IOException {
+        patterns = in.readMap(StreamInput::readString, AutoFollowPattern::new);
+        followedLeaderIndexUUIDS = in.readMapOfLists(StreamInput::readString, StreamInput::readString);
+    }
+
+    public Map<String, AutoFollowPattern> getPatterns() {
+        return patterns;
+    }
+
+    public Map<String, List<String>> getFollowedLeaderIndexUUIDS() {
+        return followedLeaderIndexUUIDS;
+    }
+
+    @Override
+    public EnumSet<MetaData.XContentContext> context() {
+        // TODO: When a snapshot is restored do we want to restore this?
+        // (Otherwise we could start following indices automatically immediately)
+        return MetaData.ALL_CONTEXTS;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return TYPE;
+    }
+
+    @Override
+    public Version getMinimalSupportedVersion() {
+        return Version.CURRENT.minimumCompatibilityVersion();
```

Comment:
This is not right, this should be 6.5.0.

Reply:
whoops